R/types-chr.R

```diff
@@ -0,0 +1,295 @@
+#' # a conversion. R will assume that the string is encoded according
+#' # to the current locale:
+#' \dontrun{
+#' bad <- set_names(str_set_encoding(latin1, "unknown"))
+#' set_utf8_locale()
+#'
+#' str_encoding(names(c(bad)))
+#' str_bytes(names(c(bad)))
+#' }
+chr_set_encoding <- function(x, encoding = c("unknown", "UTF-8", "latin1", "bytes")) {
+  Encoding(x) <- match.arg(encoding)
+  x
+}
+#' @rdname chr_set_encoding
+#' @export
+chr_encoding <- function(x) {
+  Encoding(x)
+}
+#' @rdname chr_set_encoding
+#' @export
+str_set_encoding <- function(x, encoding = c("unknown", "UTF-8", "latin1", "bytes")) {
+  stopifnot(is_string(x))
+  chr_set_encoding(x, encoding)
+}
+#' @rdname chr_set_encoding
+#' @export
+str_encoding <- function(x) {
+  stopifnot(is_string(x))
+  Encoding(x)
+}
+
+#' Set the locale's codeset for testing.
+#'
+#' Setting a locale's codeset (specifically, the \code{LC_CTYPE}
+#' category) produces side effects in R's handling of strings. The
+#' most important of these affects how the R parser marks strings. R
+#' has specific internal support for latin1 (single-byte encoding) and
+#' UTF-8 (multi-bytes variable-width encoding) strings. If the locale
+#' codeset is latin1 or UTF-8, the parser will mark all strings with
+#' the corresponding encoding. It is important for strings to have
+#' consistent encoding markers, as they determine a number of side
+#' effects in the handling of strings (see
+#' \code{\link{str_set_encoding}()}).
+#'
+#' UTF-8 is the only variable-width encoding supported in R.
+#' Multi-byte fixed-width encodings like UCS-2 are supported by R, but
+#' supplying UTF-16 strings (the default Unicode encoding on Windows)
+#' will cause the parser to choke on the character delimiters (though
+#' you may not see that behaviour in Windows IDEs if they translate
+#' inputs to another encoding for you). Note finally that
+#' \code{set_utf8_locale()} will not work on Windows as only latin1
+#' and MBCS locales are supported on this OS.
+#'
+#' Note that these helpers are only provided for testing interactively
+#' the effects of changing locale codeset. They let you quickly change
+#' the default text encoding to latin1, UTF-8, or non-UTF-8 MBCS. They
+#' are not widely tested and do not provide a way of setting the
+#' language and region of the locale. They have permanent side effects
+#' and should not be used in package examples, unit tests, or in the
+#' course of a data analysis.
+#'
+#' @export
+set_utf8_locale <- function() {
+  if (.Platform$OS.type == "windows") {
+    warn("UTF-8 is not supported on Windows")
+  } else {
+    cat("Locale codeset is now UTF-8\n")
+    invisible(Sys.setlocale("LC_CTYPE", locale = "en_US.UTF-8"))
+  }
+}
+#' @rdname set_utf8_locale
+#' @export
+set_latin1_locale <- function() {
+  if (.Platform$OS.type == "windows") {
+    locale <- "English_United States.1252"
+  } else {
+    locale <- "en_US.ISO8859-1"
```

Comment:
This is `"en_US.iso88591"` on my Ubuntu system. Does it make sense to use a system call to e.g. `locale -a` to find the right locale?